src/backends/policy_engine/souffle/check_policy_compliance.cc

```diff
@@ -69,29 +72,60 @@ int main(int argc, char* argv[]) {
   absl::SetProgramUsageMessage(kUsageMessage);
   absl::ParseCommandLine(argc, argv);
 
+  // Read the sql policy rules file.
+  absl::StatusOr<std::string> sql_policy_rules =
+          ReadFileContents(absl::GetFlag(FLAGS_sql_policy_rules));
+  if (!sql_policy_rules.ok()) {
+      LOG(ERROR) << "Error reading sql policy rules file: "
+                 << sql_policy_rules.status();
+      return 2;
+  }
+
+  // TODO(cromwellian): these flags are mutually exclusive, add a check
   // Parse the given IR file.
-  absl::StatusOr<std::string> ir_string =
-      ReadFileContents(absl::GetFlag(FLAGS_ir));
-  if (!ir_string.ok()) {
-    LOG(ERROR) << "Error reading IR file: " << ir_string.status();
-    return 2;
+  auto ir_path = absl::GetFlag(FLAGS_ir);
+  auto proto_path = absl::GetFlag(FLAGS_proto);
+
+  if (ir_path.has_value() && proto_path.has_value()) {
+      LOG(ERROR) << "Both --ir and --proto cannot be used simultaneously.";
+      return 2;
   }
 
-  // Read the sql policy rules file.
-  absl::StatusOr<std::string> sql_policy_rules =
-      ReadFileContents(absl::GetFlag(FLAGS_sql_policy_rules));
-  if (!ir_string.ok()) {
-    LOG(ERROR) << "Error reading sql policy rules file: "
-               << sql_policy_rules.status();
-    return 2;
+  std::unique_ptr<raksha::ir::Module> ir_module;
+  std::unique_ptr<raksha::ir::SsaNames> ir_ssa_names;
+  std::unique_ptr<raksha::ir::IRContext> ir_context;
+
+  if (ir_path.has_value()) {
+      absl::StatusOr<std::string> ir_string = ReadFileContents(*ir_path);
+      if (!ir_string.ok()) {
+          LOG(ERROR) << "Error reading IR file: " << ir_string.status();
+          return 2;
+      }
+
+      IrProgramParser irParser;
+      auto [context, module, ssa_names] = irParser.ParseProgram(*ir_string);
+      ir_module = std::move(module);
+      ir_ssa_names = std::move(ssa_names);
+      ir_context = std::move(context);
+  }
+
+  if (proto_path.has_value()) {
+      absl::StatusOr<std::string> proto_string = ReadFileContents(*proto_path);
+      if (!proto_string.ok()) {
+          LOG(ERROR) << "Error reading proto file: " << proto_string.status();
+          return 2;
+      }
+      raksha::ir::IRContext irContext;
+      auto [module, ssa_names] = raksha::ir::ProtoToIR::Convert(irContext, proto_string.value());
+      ir_module = std::move(module);
+      ir_ssa_names = std::move(ssa_names);
   }
 
   // Invoke policy checker and return result.
   SoufflePolicyChecker checker;
-  IrProgramParser irParser;
-  auto [context, module, ssa_names] = irParser.ParseProgram(*ir_string);
   Policy policy(*sql_policy_rules);
-  if (checker.IsModulePolicyCompliant(*module, policy)) {
+
```

Comment:
Probably should check that the `ir_*` `unique_ptrs` are all initialized here to avoid a potential seg fault with no error message.